pkg/extensions/tbtc/tbtc.go

```diff
@@ -97,6 +149,144 @@ func (t *tbtc) monitorRetrievePubKey(
 	return nil
 }
 
+func (t *tbtc) monitorProvideRedemptionSignature(
+	ctx context.Context,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) error {
+	monitoringStartFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		return t.chain.OnDepositRedemptionRequested(
+			func(
+				depositAddress string,
+				requesterAddress string,
+				digest [32]uint8,
+				utxoValue *big.Int,
+				redeemerOutputScript []uint8,
+				requestedFee *big.Int,
+				outpoint []uint8,
+				blockNumber uint64,
+			) {
+				event := &depositRedemptionRequestedEvent{
+					&baseDepositEvent{depositAddress},
+					digest,
+					blockNumber,
+				}
+
+				// Log the event to make it available for future provide
+				// redemption proof monitoring process.
+				t.eventsLog.logDepositRedemptionRequestedEvent(
+					depositAddress,
+					event,
+				)
+
+				handler(event)
+			},
+		)
+	}
+
+	monitoringStopFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		signatureSubscription, err := t.chain.OnDepositGotRedemptionSignature(
+			func(depositAddress string) {
+				handler(&baseDepositEvent{depositAddress})
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		redeemedSubscription, err := t.chain.OnDepositRedeemed(
+			func(depositAddress string) {
+				handler(&baseDepositEvent{depositAddress})
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return subscription.NewEventSubscription(
+			func() {
+				signatureSubscription.Unsubscribe()
+				redeemedSubscription.Unsubscribe()
+			},
+		), nil
+	}
+
+	actFn := func(monitoringContext *depositMonitoringContext) error {
+		depositRedemptionRequestedEvent, ok := monitoringContext.startEvent.(*depositRedemptionRequestedEvent)
+		if !ok {
+			return fmt.Errorf(
+				"monitoring context contains unexpected type of start event",
+			)
+		}
+
+		depositAddress := depositRedemptionRequestedEvent.depositAddress
+
+		keepAddress, err := t.chain.KeepAddress(depositAddress)
+		if err != nil {
+			return err
+		}
+
+		signatureSubmittedEvents, err := t.chain.PastSignatureSubmittedEvents(
+			keepAddress,
+			depositRedemptionRequestedEvent.blockNumber,
+		)
+		if err != nil {
+			return err
+		}
+
+		depositDigest := depositRedemptionRequestedEvent.digest
+
+		for _, signatureSubmittedEvent := range signatureSubmittedEvents {
+			if bytes.Equal(signatureSubmittedEvent.Digest[:], depositDigest[:]) {
+				// We add 27 to the recovery ID to align it with ethereum and
+				// bitcoin protocols where 27 is added to recovery ID to
+				// indicate usage of uncompressed public keys.
+				v := 27 + signatureSubmittedEvent.RecoveryID
+
+				return t.chain.ProvideRedemptionSignature(
+					depositAddress,
+					v,
+					signatureSubmittedEvent.R,
+					signatureSubmittedEvent.S,
+				)
+			}
+		}
+
+		return fmt.Errorf(
+			"could not find signature for digest: [%v]",
```

Comment:
Worth confirming at some point if the digest is logged in a human-readable form this way.